src/super_context.ml

```diff
@@ -865,53 +865,120 @@ module PP = struct
           mli) in
     { m with impl ; intf }
 
+  (* laziness is needed to support the case where a user is preprocessing a
+     reason file with action but linting with ppx. *)
+  let lint_module sctx ~(source : Module.t) ~(ast : Module.t Lazy.t) ~dir
+        ~dep_kind ~lint ~lib_name ~scope =
+    let alias = Alias.lint ~dir in
+    match Preprocess_map.find source.name lint with
+    | No_preprocessing -> ()
+    | Action action ->
+      let action = Action.U.Chdir (root_var, action) in
+      Module.iter source ~f:(fun _ (src : Module.File.t) ->
+        let digest_path =
+          Alias.add_action_dep
+            ~action:(Some action)
+            ~action_deps:[Dep_conf.File (String_with_vars.virt __POS__ src.name)]
+            (aliases sctx) alias in
+        let src = Path.relative dir src.name in
+        add_rule sctx
+          (Build.path src
+           >>^ (fun _ -> [src])
+           >>>
+           Build.progn
+             [ Action.run sctx
+                 action
+                 ~dir
+                 ~dep_kind:Required
+                 ~targets:(Static [])
+                 ~scope
+             ; Build.create_file digest_path
+             ])
+      )
+    | Pps { pps; flags } ->
+      let ppx_exe = get_ppx_driver sctx pps ~dir ~dep_kind in
+      Module.iter (Lazy.force ast) ~f:(fun kind src ->
+        let src_path = Path.relative dir src.name in
+        let args =
+          [ Arg_spec.As flags
+          ; As (cookie_library_name lib_name)
+          ; Ml_kind.ppx_driver_flag kind
+          ; Dep src_path
+            (* ; A "--null" *)
+          ] in
+        let digest_path =
+          Alias.add_action_dep (aliases sctx) alias
+            ~action_deps:[Dep_conf.File (String_with_vars.virt __POS__ src.name)]
+            ~action:(
+              let (args, paths) = Arg_spec.expand ~dir args () in
+              Some (Action.U.Run (
+                String_with_vars.virt __POS__ (Path.to_string ppx_exe),
+                List.map ~f:(String_with_vars.virt __POS__)
+                  (args @ (List.map ~f:Path.to_string (Pset.elements paths)))
+              ))
```

Comment:
It's a bit unfortunate to have all this code just to compute a digest. What about adding a function like this in `Alias` instead?

```ocaml
val add_stamp_dep : t -> data:Sexp.t -> Path.t
```

Reply:
That's how it it was before and it wasn't any shorter to generate the sexp. It also made the other calls longer. I don't have a strong opinion either way.